scripts/zfs-tests.sh

```diff
@@ -152,6 +153,82 @@ find_runfile() {
 }
 
 #
+# Symlink file if it appears under any of the given paths.
+#
+create_links() {
+	local dir_list="$1"
+	local file_list="$2"
+
+	[ -n $STF_PATH ] || fail "STF_PATH wasn't correctly set"
+
+	for i in $file_list; do
+		for j in $dir_list; do
+			[ ! -e "$STF_PATH/$i" ] || continue
+
+			if [ ! -d "$j/$i" -a -e "$j/$i" ]; then
+				ln -s $j/$i $STF_PATH/$i || \
+				    fail "Couldn't link $i"
+				break
+			fi
+		done
+
+		[ ! -e $STF_PATH/$i ] && STF_MISSING_BIN="$STF_MISSING_BIN$i "
+	done
+}
+
+#
+# Constrain the path to limit the available binaries to a known set.
+# When running in-tree a top level ./bin/ directory is created for
+# convenience, otherwise a temporary directory is used.
+#
+constrain_path() {
+	. $STF_SUITE/include/commands.cfg
+
+	if [ -n "${INTREE}" ]; then
+		STF_PATH="$BUILDDIR/bin"
+		if [ ! -d "$STF_PATH" ]; then
+			mkdir "$STF_PATH"
+		fi
+	else
+		SYSTEMDIR=${SYSTEMDIR:-/var/tmp/constrained_path.XXXX}
+		STF_PATH=$(/bin/mktemp -d "$SYSTEMDIR")
+	fi
+
+	STF_MISSING_BIN=""
+	chmod 755 $STF_PATH || fail "Couldn't chmod $STF_PATH"
+
+	# Standard system utilities
+	create_links "/bin /usr/bin /sbin /usr/sbin" "$SYSTEM_FILES"
+
+	if [ -z "${INTREE}" ]; then
+		# Special case links for standard zfs utilities
+		create_links "/bin /usr/bin /sbin /usr/sbin" "$ZFS_FILES"
+
+		# Special case links for zfs test suite utilties
+		create_links "$TESTSDIR/bin" "$ZFSTEST_FILES"
+	else
+		# Special case links for standard zfs utilities
+		DIRS="$(find "$CMDDIR" -type d \( ! -name .deps -a \
+		    ! -name .libs \) -print | tr '\n' ' ')"
+		create_links "$DIRS" "$ZFS_FILES"
+
+		# Special case links for zfs test suite utilties
+		DIRS="$(find "$TESTSDIR" -type d \( ! -name .deps -a \
+		    ! -name .libs \) -print | tr '\n' ' ')"
+		create_links "$DIRS" "$ZFSTEST_FILES"
+	fi
+
+	# Exceptions
+	ln -fs $STF_PATH/awk $STF_PATH/nawk
+	ln -fs $STF_PATH/gzip $STF_PATH/compress
+	ln -fs $STF_PATH/parted $STF_PATH/format
```

Comment:
We shouldn't link `format` to `parted` since they take such different command line options.  Better for it to fail outright as an unknown command.  These commands can only be correctly used on a specific platform.